File: src/Y2024/Day21/main.py
from collections.abc import Sequence, Iterable
from itertools import pairwise
from typing import TextIO
from functools import cache

from Tools.Python.path_stuff import *


def parser(raw_data: TextIO):
    return raw_data.read().splitlines()


def make_grid(keypad):
    res = {
        x + 1j*y: v
        for x, row in enumerate(keypad.splitlines())
        for y, v in enumerate(row)
        if v != ' '
    }
    return res


NUM_KEYPAD = make_grid("""
789
456
123
 0A
""".strip('\n'))
INV_NUM_KEYPAD = {v: k for k, v in NUM_KEYPAD.items()}

DIR_KEYPAD = make_grid("""
 ^A
<v>
""".strip('\n'))
INV_DIR_KEYPAD = {v: k for k, v in DIR_KEYPAD.items()}
a_key = INV_DIR_KEYPAD['A']


def sgn(x: float) -> int:
    return (x > 0) - (x < 0)


def get_dirs(d: complex) -> list[complex]:
    d1, d2 = sgn(d.real), sgn(d.imag)
    if d1: yield d1
    if d2: yield d2 * 1j


d_dict = {
    d: INV_DIR_KEYPAD[s]
    for d, s in zip((-1, 1j, 1, -1j), '^>v<')
}


@cache
def all_paths(start: complex, end: complex, is_num: bool = False) -> Iterable[tuple[complex, ...]]:
    grid = NUM_KEYPAD if is_num else DIR_KEYPAD

    def solve(loc: complex):
        if loc == end:
            yield ()
            return
        for d in get_dirs(end - loc):
            if loc + d in grid:
                for p in solve(loc + d):
                    yield d_dict[d], *p
    return [*solve(start)]


def minimize_seq(keys: tuple[complex, ...], n: int, is_first: bool = False) -> int:
    if not n:
        return len(keys) - 1
    return sum(
        minimize_key(start, end, n - 1, is_first)
        for start, end in pairwise(keys)
    )


@cache
def minimize_key(start: complex, end: complex, n: int, is_first: bool) -> int:
    return min(
        minimize_seq((a_key, *path, a_key), n)
        for path in all_paths(start, end, is_first)
    )


def solve(data: list[str], num_keypads: int) -> int:
    return sum(
        minimize_seq(
            tuple(INV_NUM_KEYPAD[c] for c in f'A{code}'),
            num_keypads,
            True,
        ) * int(code[:-1])
        for code in data
    )


def full_solver(data: list[str]):
    return solve(data, 3), solve(data, 26)


if __name__ == '__main__':
    testing = False

    with open(test_path if testing else 'input.txt', 'r') as file:
        data = parser(file)
    print(*full_solver(data))

File: src/Y2024/Day21/main.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Numerics;

static var keypad = new Dictionary<char, Complex> {
    {'7', new Complex(-3, -2)},
    {'8', new Complex(-3, -1)},
    {'9', new Complex(-3, 0)},
    {'4', new Complex(-2, -2)},
    {'5', new Complex(-2, -1)},
    {'6', new Complex(-2, 0)},
    {'1', new Complex(-1, -2)},
    {'2', new Complex(-1, -1)},
    {'3', new Complex(-1, 0)},
    {'0', new Complex(0, -1)},
    {'A', new Complex(0, 0)},
    {'^', new Complex(0, -1)},
    {'<', new Complex(1, -2)},
    {'v', new Complex(1, -1)},
    {'>', new Complex(1, 0)},
};

static (List<Complex>, int)[] parse(string filename) {
    string[] lines = File.ReadAllLines(filename);
    var rtn = new (List<Complex>, int)[lines.Length];
    for (int i = 0; i < lines.Length; i++) {
        string line = lines[i];
        var seq = new List<Complex> {keypad['A']};
        foreach (var c in line) {
            seq.Add(keypad[c]);
        }
        int num;
        int.TryParse(line.TrimEnd('A'), out num);
        rtn[i] = (seq, num); 
    }
    return rtn;
}

static var paths_cache = new Dictionary<(Complex, Complex), List<List<Complex>>>();
static var d_dict = new Dictionary<Complex, Complex> {
    {new Complex(-1, 0), new Complex(0, -1)},
    {new Complex(0, 1), new Complex(1, 0)},
    {new Complex(1, 0), new Complex(1, -1)},
    {new Complex(0, -1), new Complex(1, -2)},
};

static List<List<Complex>> find_all_paths(Complex start, Complex end) {
    if (paths_cache.ContainsKey((start, end))) {
        return paths_cache[(start, end)];
    }
    var paths = new List<List<Complex>>();
    void search(Complex z, List<Complex> path) {
        if (z == new Complex(0, -2)) {return;}
        if (z == end) {
            paths.Add(path.ToList());
            return;
        }
        Complex d = end - z;
        if (d.Real != 0) {
            Complex delta = new Complex(double.Sign(d.Real), 0);
            path.Add(d_dict[delta]);
            search(z + delta, path);
            path.RemoveAt(path.Count - 1);
        }
        if (d.Imaginary != 0) {
            Complex delta = new Complex(0, double.Sign(d.Imaginary));
            path.Add(d_dict[delta]);
            search(z + delta, path);
            path.RemoveAt(path.Count - 1);
        }
    }
    search(start, new List<Complex>());
    paths_cache[(start, end)] = paths;
    return paths;
}

static long minimize_seq(List<Complex> keys, int n, bool is_first = false) {
    if (n == 0) {return keys.Count - 1;}
    long sum = 0;
    for (int i = 0; i < keys.Count - 1; i++) {
        var start = keys[i];
        var end = keys[i + 1];
        sum += minimize_key(start, end, n - 1, is_first);
    }
    return sum;
}

static var key_cache = new Dictionary<(Complex, Complex, int), long>();

static long minimize_key(Complex start, Complex end, int n, bool is_first) {
    var triple = (start, end, n);
    if (!is_first && key_cache.ContainsKey(triple)) {
        return key_cache[triple];
    }
    long min = -1;
    foreach (var path in find_all_paths(start, end)) {
        var seq = new List<Complex> {new Complex(0, 0)};
        seq.AddRange(path);
        seq.Add(new Complex(0, 0));
        var res = minimize_seq(seq, n);
        if (min == -1 || res < min) {
            min = res;
        }
    }
    if (!is_first) {
        key_cache[triple] = min;
    }
    return min;
}

static long solve((List<Complex>, int)[] data, int num_keypads) {
    long sum = 0;
    foreach (var (seq, num) in data) {
        sum += minimize_seq(seq, num_keypads, true) * num;
    }
    return sum;
}

var data = parse("input.txt");
Console.WriteLine(solve(data, 3));
Console.WriteLine(solve(data, 26));

